### PR TITLE
Add yolov5n to tests

### PR DESCRIPTION
**Pull request by skycat8**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ X] I agree to contribute to the project under Apache 2 License.
- [ X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ X] The PR is proposed to the proper branch
- [ X] There is a reference to the original bug report and related work
- [ X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ X] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by Abdurrahheem** [comment]
@skycat8 could you share the onnx weights of the model please?

File: modules/dnn/test/test_onnx_importer.cpp [hunk 1]
```diff
@@ -2669,6 +2669,89 @@ TEST_P(Test_ONNX_layers, YOLOv7)
     normAssertDetections(refClassIds, refScores, refBoxes, keep_classIds, keep_confidences, keep_boxes);
 }
 
+
+TEST_P(Test_ONNX_layers, YOLOv5m)
+{
+    std::string weightPath = _tf("models/yolov5m.onnx", false);
```
Comment (asmorkalov) [inline modules/dnn/test/test_onnx_importer.cpp:2675]: You need to create PR with the model link in opencv_extra. PR should be created with the same branch name. Test models are downloaded by this script and listedt inside of it: https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/download_models.py

File: modules/dnn/test/test_onnx_importer.cpp [hunk 2]
```diff
@@ -2669,6 +2669,89 @@ TEST_P(Test_ONNX_layers, YOLOv7)
...
+    std::vector<float> keep_confidences;
+    std::vector<Rect2d> keep_boxes;
+    for (auto i : keep_idx)
+    {
+        keep_classIds.push_back(classIds[i]);
+        keep_confidences.push_back(confidences[i]);
+        keep_boxes.push_back(boxes[i]);
+    }
+    for (int i = 0; i <= 2; i++) 
+        cout << keep_classIds[i];
```
Comment (asmorkalov) [inline modules/dnn/test/test_onnx_importer.cpp:2749]: Please remove it from test.

File: modules/dnn/test/test_onnx_importer.cpp (hunk 1 above)
Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:2675]: And there is already YOLOv5n, no need for 5m version: https://github.com/opencv/opencv/blob/0e151e3c8837441c961d7f4788c733d634cd4964/modules/dnn/perf/perf_net.cpp#L275

Comment (Abdurrahheem) [reply modules/dnn/test/test_onnx_importer.cpp:2675]: @are they the same though?

Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:2675]: I think they differ in number of repeated blocks or something.

Comment (Abdurrahheem) [reply modules/dnn/test/test_onnx_importer.cpp:2675]: @dkurt  Ar you suggesting to use weights of yolox5n  for this test or to remove the test entirely?

Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:2675]: Accuracy test is a good addition. For now there is only performance test for YOLOv5 as I see. So I suggest replace YOLOv5m to YOLOv5n.

File: modules/dnn/test/test_onnx_importer.cpp (hunk 2 above)
Comment (skycat8) [reply modules/dnn/test/test_onnx_importer.cpp:2749]: @dkurt As you recommended, I checked yolov5n:

1. On my test image it works worse than yolov5m: bicycle box is predicted incorrectly.
2. Opencv result detection result on my test image differs from ultralytics result: scores are lower

I attach ultralitics detection visualization and my visualization of the same model (yolov5n.onnx) for the same image.

In [ultralytis](https://github.com/ultralytics/yolov5) authors also report better performance of yolov5m in comparison with yolov5n.

Should I still add a test for yolov5n? Or should I add  append opencv_extra with yolov5m link?



ultralytics_result
opencv_result

Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:2749]: Let’s start from the motivation. If both 5n and 5m models consist of the same layers and there is no difference in the testing perspective which one to test - use 5n. Tests are needed to cover OpenCV implementation. For demonstration object_detection.py sample should be used.

So if you can, please keep test run 5n but put a link to yopo5m to the sample as it’s done here: https://github.com/opencv/opencv/pull/24496

File: modules/dnn/test/test_onnx_importer.cpp [hunk 3]
```diff
@@ -2669,6 +2669,87 @@ TEST_P(Test_ONNX_layers, YOLOv7)
...
+
+    Mat img = imread(imgPath);
+    Mat inp = blobFromImage(img, 1/255.0, targetSize, Scalar(0, 0, 0), true, false);
+
+    Net net = readNet(weightPath);
+
+    net.setInput(inp);
+    std::vector<Mat> outs;
+    net.forward(outs, net.getUnconnectedOutLayersNames());
+    Mat preds = outs[0].reshape(1, outs[0].size[1]); // [1, 25200, 85]
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:2699]: `Mat preds = net.forward()`

File: modules/dnn/test/test_onnx_importer.cpp [hunk 4]
```diff
@@ -2669,6 +2669,86 @@ TEST_P(Test_ONNX_layers, YOLOv7)
     normAssertDetections(refClassIds, refScores, refBoxes, keep_classIds, keep_confidences, keep_boxes);
 }
 
+
+TEST_P(Test_ONNX_layers, YOLOv5n)
+{
+    std::string weightPath = _tf("models/yolov5n.onnx", false);
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:2675]:
```suggestion
    std::string weightPath = findDataFile("dnn/yolov5n.onnx", false);
```

File: modules/dnn/test/test_onnx_importer.cpp [hunk 5]
```diff
@@ -2669,6 +2669,86 @@ TEST_P(Test_ONNX_layers, YOLOv7)
...
+
+        // get bbox coords
+        float* det = preds.ptr<float>(i);
+        double cx = det[0];
+        double cy = det[1];
+        double w = det[2];
+        double h = det[3];
+        // [x1, y1, x2, y2]
+        boxes.push_back(Rect2d(cx - 0.5 * w, cy - 0.5 * h,
+                                cx + 0.5 * w, cy + 0.5 * h));
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:2729]: alignment

File: modules/dnn/test/test_onnx_importer.cpp [hunk 6]
```diff
@@ -2669,6 +2669,86 @@ TEST_P(Test_ONNX_layers, YOLOv7)
...
+            continue;
+
+        // get bbox coords
+        float* det = preds.ptr<float>(i);
+        double cx = det[0];
+        double cy = det[1];
+        double w = det[2];
+        double h = det[3];
+        // [x1, y1, x2, y2]
+        boxes.push_back(Rect2d(cx - 0.5 * w, cy - 0.5 * h,
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:2728]: Trailing whitespace

File: modules/dnn/test/test_onnx_importer.cpp (hunk 4 above)
Comment (Abdurrahheem) [reply modules/dnn/test/test_onnx_importer.cpp:2675]: There is a quite a number of test with `_tf` as reader. Which one to use in general ?

Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:2675]: They are locally defined methods to keep paths shorter. However this model is located in a different folder:
https://github.com/opencv/opencv/blob/7783c229771259e16e70c047af074c29c66eb324/modules/dnn/test/test_onnx_importer.cpp#L15C1-L20C1

File: modules/dnn/test/test_onnx_importer.cpp [hunk 7]
```diff
@@ -2669,6 +2669,86 @@ TEST_P(Test_ONNX_layers, YOLOv7)
     normAssertDetections(refClassIds, refScores, refBoxes, keep_classIds, keep_confidences, keep_boxes);
 }
 
+
+TEST_P(Test_ONNX_layers, YOLOv5n)
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:2673]: Wrong fixture. Use `Test_ONNX_nets`: `TEST_P(Test_ONNX_nets, YOLOv5n)`

Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:2673]: Please also fix for `TEST_P(Test_ONNX_layers, YOLOv7)` -> `TEST_P(Test_ONNX_nets, YOLOv7)`

File: modules/dnn/test/test_onnx_importer.cpp [hunk 8]
```diff
@@ -2669,6 +2669,86 @@ TEST_P(Test_ONNX_layers, YOLOv7)
...
+    std::vector<Rect2d> keep_boxes;
+    for (auto i : keep_idx)
+    {
+        keep_classIds.push_back(classIds[i]);
+        keep_confidences.push_back(confidences[i]);
+        keep_boxes.push_back(boxes[i]);
+    }
+
+    normAssertDetections(refClassIds, refScores, refBoxes, keep_classIds, keep_confidences, keep_boxes);
+}
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:2749]: There is a duplicated code with `YOLOv7`. I suggest add a method such as `testYOLO` which accepts a path to tested model and reference predictions.

Comment (skycat8) [reply modules/dnn/test/test_onnx_importer.cpp:2749]: YOLOv7 inference return std::vector<Mat>, and it is requited to take 3rd element and reshape it for further postprocessing;
YOLOv5n inference return just Mat object.

We can:
1. replace YOLOv7 onnx file to make t return only one object
2. write common method which accepts network outputs (direct outputs for yolov5n and processed outputs for yolov7) and reference predictions

Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:2691]: I have pushed a commit with my view. Let's see if tests pass

**Review by dkurt — APPROVED** [review]

**Pull request merged**